File: CNBlurredAbbreviatedLabel.m
//
//  CNBlurredAbbreviatedLabel.m
//  CircleNameLabel
//
//  Created by Adi Mathew on 4/27/16.
//  Copyright © 2016 RCPD. All rights reserved.
//

#import "CNBlurredAbbreviatedLabel.h"

@interface CNBlurredAbbreviatedLabel()

@property (nonatomic, strong) UILabel *letterLabel;
@property (nonatomic, strong) UIVisualEffectView *blurView;
@property (nonatomic, strong) UIVisualEffectView *vibView;

@end

@implementation CNBlurredAbbreviatedLabel

-(void)setLetters:(NSString *)letters {
    _letters = letters;
    [UIView animateWithDuration:0.5 animations:^{
        _letterLabel.text = letters; //CHECK FONT HERE
        CGFloat size = idealFontSizeForFont([UIFont systemFontOfSize:100.0f], self.bounds, self.letters);
        _letterLabel.font = [UIFont systemFontOfSize: size];
    }];
}

- (void)setColor:(UIColor *)color {
    _color = color;
    if (color != [UIColor lightGrayColor]) {
        _blurView.contentView.backgroundColor = [color colorWithAlphaComponent:_colorAlpha - 0.1];
    } else {
        _blurView.contentView.backgroundColor = [color colorWithAlphaComponent:_colorAlpha];
    }

}

-(void)setColorAlpha:(CGFloat)colorAlpha {
    _colorAlpha = colorAlpha;
    _blurView.contentView.backgroundColor = [_color colorWithAlphaComponent: colorAlpha];
}

-(void)setLabelRadiusFactor:(CGFloat)labelRadiusFactor {
    _labelRadiusFactor = labelRadiusFactor;
    self.layer.cornerRadius = self.frame.size.width/labelRadiusFactor;
}

-(instancetype)initWithFrame:(CGRect)frame {
    return [self initWithFrame:frame text:@"."];
}

-(instancetype)initWithFrame:(CGRect)frame text: (NSString *)letters {
    NSParameterAssert(letters);
    self = [super initWithFrame:frame];
    if (self) {
        [self setDefaults];
        self.letters = letters ? letters : self.letters;
    }
    
    return  self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    return [self initWithCoder:coder text:@"."];
}

- (instancetype)initWithCoder:(NSCoder *)coder text: (NSString *)letters {
    self = [super initWithCoder:coder];
    if (self) {
        [self setDefaults];
        self.letters = letters ? letters : self.letters;
    }
    return self;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [self setDefaults];
    }
    return self;
}

-(void)layoutSubviews {
    self.blurView.frame = self.bounds;
    self.vibView.frame = self.blurView.bounds;
}

- (void)setDefaults {
    _color = [UIColor lightGrayColor];
    _colorAlpha = 0.25;
    _labelRadiusFactor = 2.0;
    
//  WTF http://stackoverflow.com/questions/28798269/round-uivisualeffectview
   
    UIBlurEffect *blur = [UIBlurEffect effectWithStyle: UIBlurEffectStyleLight];
    UIVisualEffectView *blurView = [[UIVisualEffectView alloc] initWithEffect:blur];
    blurView.contentView.backgroundColor = [self.color colorWithAlphaComponent:self.colorAlpha]; //FIXED
    
    //Changing this line is what used to cause the blurring to act weird.
    self.layer.cornerRadius = self.frame.size.width/_labelRadiusFactor;
    
    self.clipsToBounds = YES;
    [self addSubview:blurView];
    
    CGFloat size = idealFontSizeForFont([UIFont systemFontOfSize:100.0f], self.bounds, self.letters);
    size = size == NAN ? size : 60;
    
    UILabel *tx = [[UILabel alloc] initWithFrame:self.bounds];
    tx.font = [UIFont systemFontOfSize:size];
    tx.textColor = [UIColor whiteColor];
//    tx.text = self.letters ? self.letters : @"_";
    tx.textAlignment = NSTextAlignmentCenter;
    tx.backgroundColor = [UIColor clearColor];
    
    UIVibrancyEffect *vib = [UIVibrancyEffect effectForBlurEffect:blur];
    UIVisualEffectView *vibView = [[UIVisualEffectView alloc] initWithEffect:vib];
    [blurView.contentView addSubview:vibView];
    [vibView.contentView addSubview:tx];
    
    self.letterLabel = tx;
    self.blurView = blurView;
    self.vibView = vibView;
}

static CGFloat idealFontSizeForFont( UIFont *font, CGRect rect, NSString *text) {
    CGFloat ppp = font.lineHeight/font.pointSize;
    CGFloat idealPointSize = rect.size.height * ppp;
    if (text.length == 1) {
        idealPointSize = 0.6 * idealPointSize;
    } else {
        idealPointSize = 0.75 * idealPointSize / text.length;
    }
    
    return idealPointSize;
}


@end

File: CircleNameLabel/CircleNameLabel/CNLabelTest.playground/Sources/CNBlurredAbbreviatedLabel.swift
import UIKit

public class CNBlurredAbbreviatedLabel: UIView {
    public var letters = "_" {
        willSet(newLetters) {
            letterLabel.text = newLetters
            let size = idealSize(UIFont.systemFontOfSize(100.0), self.bounds, newLetters)
            letterLabel.font = UIFont.systemFontOfSize(size)
        }
    }
    
    private var letterLabel = UILabel()
    private var blurView = UIVisualEffectView()
    private var vibView = UIVisualEffectView()
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        self.setDefaults()
    }
    
    convenience public init() {
        self.init(frame: CGRect.zero)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.setDefaults()
    }
    
    override public func layoutSubviews() {
        blurView.frame = self.bounds
        vibView.frame = self.blurView.bounds
    }
    
    func setEffectViewDefaults(frame: CGRect) {
        let blurEffect = UIBlurEffect(style: .Light)
        let vibEffect = UIVibrancyEffect(forBlurEffect: blurEffect)
        blurView.effect = blurEffect
        vibView.effect = vibEffect
        
        letterLabel.frame = frame
    }
    
    func setDefaults() {
        self.layer.cornerRadius = self.bounds.size.width/2.0
        self.clipsToBounds = true
        //baseView.backgroundColor = UIColor.whiteColor().colorWithAlphaComponent(0.25)
        
        blurView.frame = self.bounds
        blurView.contentView.backgroundColor = UIColor.whiteColor().colorWithAlphaComponent(0.5)
        
        let size: CGFloat = idealSize(UIFont.systemFontOfSize(100.0), self.bounds, letters)
        
        letterLabel = UILabel(frame: self.bounds)
        letterLabel.font = UIFont.systemFontOfSize(size)
        letterLabel.textColor = UIColor.whiteColor()
        letterLabel.textAlignment = .Center
        letterLabel.backgroundColor = UIColor.clearColor()
        letterLabel.text = letters
        
        //blurView.addSubview(text)
        self.addSubview(blurView)
        

        vibView.frame = blurView.bounds
        
        blurView.contentView.addSubview(vibView)
        vibView.contentView.addSubview(letterLabel)

    }
    
    func idealSize(font: UIFont, _ rect: CGRect, _ text: String) -> CGFloat {
        let ppp = font.lineHeight/font.pointSize
        var idealPointSize = rect.size.height * ppp
        
        if text.characters.count == 1 {
            idealPointSize *= 0.6
        } else {
            idealPointSize = 0.75 * idealPointSize /  CGFloat(text.characters.count)
        }
        
        return idealPointSize
    }

}

